packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,484 @@
+    let continuationToken: string | undefined;
+
+    do {
+      const batch: S3Asset[] = [];
+
+      while (batch.length < batchSize) {
+        const response = await s3.listObjectsV2({
+          Bucket: bucket,
+          ContinuationToken: continuationToken,
+        }).promise();
+
+        response.Contents?.forEach((obj) => {
+          const key = obj.Key ?? '';
+          const size = obj.Size ?? 0;
+          if (obj.Key) {
+            batch.push(new S3Asset(bucket, key, size));
+          }
+        });
+
+        continuationToken = response.NextContinuationToken;
+
+        if (!continuationToken) break; // No more objects to fetch
+      }
+
+      if (batch.length > 0) {
+        yield batch;
+      }
+    } while (continuationToken);
+  }
+
+  /**
+   * Fetches all relevant stack templates from CloudFormation. It ignores the following stacks:
+   * - stacks in DELETE_COMPLETE or DELET_IN_PROGRES stage
+   * - stacks that are using a different bootstrap qualifier
+   *
+   * It fails on the following stacks because we cannot get the template and therefore have an imcomplete
+   * understanding of what assets are being used.
+   * - stacks in REVIEW_IN_PROGRESS stage
+   */
+  private async fetchAllStackTemplates(cfn: CloudFormation, qualifier?: string) {
+    const stackNames: string[] = [];
+    await paginateSdkCall(async (nextToken) => {
+      let response = await cfn.listStacks({ NextToken: nextToken }).promise();
+
+      // We cannot operate on REVIEW_IN_PROGRESS stacks because we do not know what the template looks like in this case
+      // If we encounter this status, we will wait up to a minute for it to land before erroring out.
+      const reviewInProgressStacks = response.StackSummaries?.filter(s => s.StackStatus === 'REVIEW_IN_PROGRESS') ?? [];
+      if (reviewInProgressStacks.length > 0) {
+        const updatedRequest = await this.waitForStacksAndRefetch(cfn, response, reviewInProgressStacks);
+        response = await updatedRequest.promise();
+      }
+
+      // Deleted stacks are ignored
+      const ignoredStatues = ['CREATE_FAILED', 'DELETE_COMPLETE', 'DELETE_IN_PROGRESS', 'DELETE_FAILED'];
+      stackNames.push(
+        ...(response.StackSummaries ?? [])
+          .filter(s => !ignoredStatues.includes(s.StackStatus))
+          .map(s => s.StackId ?? s.StackName),
+      );
+
+      return response.NextToken;
+    });
+
+    print(chalk.blue(`Parsing through ${stackNames.length} stacks`));
+
+    const templates: string[] = [];
+    for (const stack of stackNames) {
+      let summary;
+      summary = await cfn.getTemplateSummary({
+        StackName: stack,
+      }).promise();
+
+      // Filter out stacks that we KNOW are using a different bootstrap qualifier
+      // This is necessary because a stack under a different bootstrap could coincidentally reference the same hash
+      // and cause a false negative (cause an asset to be preserved when its isolated)
```

Comment:
Please indicate this is mostly for the integration tests' benefit.

If people have non-standard stacks where this test doesn't work, we might be conservative and not collect everything we could, but that's fine: the number of assets not collected that way is bounded by `the number of assets in a stack * the number of of stacks`, which is  (probably) a low number.

Reply:
This decision logic should most likely also be in a function, where it can be documented properly and easily found and changed.